### ISSUE TITLE
Layer Norm kernel fails for ROCm

### ISSUE THREAD

**Issue by NouamaneTazi** [microsoft/Megatron-DeepSpeed#95]
The test for the [fused layer norm kernel ](https://github.com/microsoft/Megatron-DeepSpeed/blob/238ab6df7e7561884fa2f3993d69099758105946/megatron/model/fused_layer_norm.py#L62) seems to fail for ROCm

<details>
<summary>Here's a small reproduction script:</summary>

```python
from megatron import fused_kernels
from megatron.model.fused_layer_norm import MixedFusedLayerNorm
from torch.nn import LayerNorm


from transformers import BertTokenizer
from transformers.models.bert.modeling_bert import BertModel
import transformers

transformers.logging.set_verbosity(
    transformers.logging.FATAL,
)

# Copied from https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/fused_kernels/tests/test_fused_kernels.py#L223
def test_layer_norm():
    tokenizer = BertTokenizer.from_pretrained("bert-base-cased")
    bert = BertModel.from_pretrained("bert-base-cased").cuda().half()
    test_text = (
        "Hello. How are you? I am fine thank you and you? yes Good. "
        "hi hi hi hi hi hi hi hi hi hi hi hi hi"  # 32
    )

    tokens = tokenizer(
        [test_text] * 4,
        return_tensors="pt",
    )

    # [bsz, seq_len, d_model]
    embedding_output = (
        bert.embeddings(
            input_ids=tokens["input_ids"].cuda(),
            position_ids=None,
            token_type_ids=tokens["token_type_ids"].cuda(),
            inputs_embeds=None,
            past_key_values_length=0,
        )
        .cuda()
        .half()
    )

    fused_layernorm_layer = (
        MixedFusedLayerNorm(normalized_shape=embedding_output.size(-1)).cuda().half()
    )

    torch_layernorm_layer = (
        LayerNorm(normalized_shape=embedding_output.size(-1)).cuda().half()
    )

    fused_output = fused_layernorm_layer(embedding_output)
    torch_output = torch_layernorm_layer(embedding_output)
    test_result = (fused_output - torch_output).abs()

    while test_result.dim() != 1:
        test_result = test_result.mean(dim=-1)

    diff = test_result.mean(dim=-1)

    if diff <= 1e-3:
        print(
            f"\n[Success] test_layer_norm"
            f"\n > mean_difference={diff}"
            f"\n > fused_values={fused_output[-1][-1][:5].tolist()}"
            f"\n > torch_values={torch_output[-1][-1][:5].tolist()}"
        )
    else:
        print(
            f"\n[Fail] test_layer_norm"
            f"\n > mean_difference={diff}, "
            f"\n > fused_values={fused_output[-1][-1][:5].tolist()}, "
            f"\n > torch_values={torch_output[-1][-1][:5].tolist()}"
        )

        
if __name__ == "__main__":     
    # initialize args
    from megatron.global_vars import _parse_args
    args_defaults = {'tokenizer_type': 'GPT2BPETokenizer', 'micro_batch_size':4, 'num_layers':24, 'hidden_size':1024, 'num_attention_heads':16, 'seq_length':1024, 'max_position_embeddings':1024, 'vocab_file':'gpt2-vocab.json', 'merge_file':'gpt2-merges.txt', 'train_samples':1000000, 'lr':3.0e-4, 'loss_scale':0}
    args = _parse_args(extra_args_provider=None,
                        defaults=args_defaults,
                        ignore_unknown_args=False)
    # compile kernels   
    print("Compiling kernels...")
    fused_kernels.load(args)
    print("Done compiling kernels.\n")

    print("Running test...")
    test_layer_norm()
```
</details>

Which gives on a DGX (CUDA): ✅
```
[Success] test_layer_norm
 > mean_difference=0.0
 > fused_values=[-0.0704345703125, -0.391845703125, 1.068359375, -0.32470703125, 0.0677490234375]
 > torch_values=[-0.0704345703125, -0.391845703125, 1.068359375, -0.32470703125, 0.0677490234375]
 ```

But gives on a MI250X (ROCm): ❌
```
[Fail] test_layer_norm
 > mean_difference=0.100830078125, 
 > fused_values=[-0.01995849609375, -0.298583984375, 0.966796875, -0.240234375, 0.0997314453125], 
 > torch_values=[-0.0704345703125, -0.391845703125, 1.068359375, -0.32470703125, 0.0677490234375]
 ```
 Any ideas on how to fix this?
 
 Related to https://github.com/microsoft/Megatron-DeepSpeed/pull/68
 cc @rraminen @jeffra  @tjruwase @jithunnair-amd

**Comment by jeffdaily** [comment]
I suspect there is a warpSize 32 vs 64 issue here.

**Comment by hubertlu-tw** [comment]
As we addressed this issue in https://github.com/microsoft/Megatron-DeepSpeed/pull/96, shall we close this? 
Thank you for bringing up this issue.

**Comment by NouamaneTazi** [comment]
Yes ofc! Thanks again.
Solved in #96
